internal/graph/storagewrapper.go

```diff
@@ -0,0 +1,162 @@
+	iter, err := c.OpenFGADatastore.ReadUsersetTuples(ctx, store, filter, options)
+	if err != nil {
+		return nil, err
+	}
+
+	cacheKey := fmt.Sprintf("%s/%s#%s", store, filter.Object, filter.Relation)
+	// todo(jpadilla): stable sort filter.AllowedUserTypeRestrictions
+	for _, userset := range filter.AllowedUserTypeRestrictions {
+		if _, ok := userset.GetRelationOrWildcard().(*openfgav1.RelationReference_Relation); ok {
+			cacheKey += "/" + userset.GetType() + ":%#" + userset.GetRelation()
+		}
+		if _, ok := userset.GetRelationOrWildcard().(*openfgav1.RelationReference_Wildcard); ok {
+			cacheKey += "/" + userset.GetType() + ":*"
+		}
+	}
+
+	return c.newCachedIterator(iter, cacheKey), nil
+}
+
+func (c *cachedDatastore) Read(ctx context.Context, store string, tupleKey *openfgav1.TupleKey, options storage.ReadOptions) (storage.TupleIterator, error) {
+	iter, err := c.OpenFGADatastore.Read(ctx, store, tupleKey, options)
+	if err != nil {
+		return nil, err
+	}
+	cacheKey := fmt.Sprintf("%s/%s", store, tuple.TupleKeyToString(tupleKey))
+	return c.newCachedIterator(iter, cacheKey), nil
+}
+
+func (c *cachedDatastore) newCachedIterator(iter storage.TupleIterator, key string) storage.TupleIterator {
+	cacheKey := cacheKeyFor(key)
+	cachedResp := c.cache.Get(cacheKey)
+	isCached := cachedResp != nil && !cachedResp.Expired && cachedResp.Value != nil
+	if isCached {
+		return storage.NewStaticTupleIterator(cachedResp.Value.([]*openfgav1.Tuple))
+	}
+
+	return &cachedIterator{
+		iter:         iter,
+		tks:          make([]*openfgav1.Tuple, 0, c.maxCacheSize),
+		cacheKey:     cacheKey,
+		cache:        c.cache,
+		maxCacheSize: c.maxCacheSize,
+		ttl:          c.ttl,
+	}
+}
+
+// Close closes the datastore and cleans up any residual resources.
+func (c *cachedDatastore) Close() {
+	c.cache.Stop()
+	c.OpenFGADatastore.Close()
+}
+
+type cachedIterator struct {
+	iter         storage.TupleIterator
+	tks          []*openfgav1.Tuple
+	cacheKey     string
+	cache        storage.InMemoryCache[any]
+	ttl          time.Duration
+	maxCacheSize int64
+	closeOnce    sync.Once
+}
+
+// Next see [Iterator.Next].
+func (c *cachedIterator) Next(ctx context.Context) (*openfgav1.Tuple, error) {
+	tuple, err := c.iter.Next(ctx)
+	if err != nil {
+		if errors.Is(err, storage.ErrIteratorDone) {
+			c.flush()
+		}
+		return nil, err
+	}
+
+	if len(c.tks) < int(c.maxCacheSize) {
+		c.tks = append(c.tks, tuple)
+	}
+
+	return tuple, nil
+}
+
+// Stop see [Iterator.Stop].
+func (c *cachedIterator) Stop() {
+	c.closeOnce.Do(func() {
+		// todo(jpadilla): figure out how to handle when at capacity
+		if len(c.tks) == int(c.maxCacheSize) {
+			c.iter.Stop()
+			return
+		}
+
+		go func() {
+			for {
+				_, err := c.Next(context.Background())
+				if err != nil {
+					c.flush()
+					break
+				}
+			}
+
+			c.iter.Stop()
+		}()
```

Comment:
what synchronizes this? wondering how we are preventing go routine leaks. while this operation may be benign when orphaned, we shouldn't allow orphaned routines in practice.